django/db/models/sql/where.py

```diff
@@ -37,6 +38,50 @@ class WhereNode(tree.Node):
     """
     default = AND
 
+    def split_having(self, negated=False):
+        """
+        Returns two possibly None nodes, one for those parts of self that
+        should be pushed to having, and one for those parts of self
+        that should be in where.
+        """
+        in_negated = negated ^ self.negated
+        split_pos = (
+            (in_negated and self.connector == AND) or
+            (not in_negated and self.connector == OR))
+        if split_pos and self.need_having(self):
+            return self, None
+        having_parts = []
+        where_parts = []
+        for c in self.children:
+            if isinstance(c, tree.Node):
+                having_part, where_part = c.split_having(in_negated)
+            elif self.need_having(c):
+                having_part, where_part = c, None
+            else:
+                having_part, where_part = None, c
+            if having_part:
+                having_parts.append(having_part)
+            if where_part:
+                where_parts.append(where_part)
+        if having_parts:
+            having_node = self.__class__(having_parts, self.connector, in_negated)
```

Comment:
I would write something like this to avoid the repetition:
```
having_nodes = self._class__(*(having_parts, self.connector, in_negated) if having_parts else ())
```
What do you think?